Fix broadcast of large messages

marianafranco opened the pull request:
**What this PR does**:
Reverting https://github.com/grafana/dskit/pull/85 as pods are failing to unmarshal large messages causing `failed to unmarshal received KV Pair` errors. This can be verified by increasing the number of cortex instances on the `TestSingleBinaryWithMemberlistScaling` integration test.

I tried to update cortex to include https://github.com/grafana/memberlist/pull/1 but the problem persisted.

This also seems to be related to pods getting OOM killed: https://github.com/cortexproject/cortex/issues/4668

**Which issue(s) this PR fixes**:
Fixes #4668 (still need to be confirmed).

**Checklist**
- [X] Tests updated
- [ ] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

marianafranco commented:
@pracucci Could you take a look on this one since you worked on the change that is going to be reverted? There is probably some work that need to be done on the [KeyValuePair.Unmarshal()](https://github.com/cortexproject/cortex/blame/03407c5ee5824606b06ecd6256a1067179334b17/pkg/ring/kv/memberlist/kv.pb.go#L507) method before this can be enabled again.

alvinlin123 commented:
According to https://github.com/grafana/dskit/blob/01ce9286d7d52c7ddcfd0a63ef42d240ea4454d2/go.mod#L43

I think we also need to add similar `replace` directive to Cortex's `go.mod`, so Cortex would have this change: https://github.com/hashicorp/memberlist/commit/8d2a27af7d1d23119da66585d71f7b1e60fde3d3

alvinlin123 commented:
After applying https://github.com/hashicorp/memberlist/commit/8d2a27af7d1d23119da66585d71f7b1e60fde3d3 and running `TestSingleBinaryWithMemberlistScaling` integ test, I am still seeing `ailed to unmarshal received KV Pair`

pracucci commented:
I agree the current code in Cortex is not working. There are two ways you can handle it:

1. Revert the change (as done in this PR)
2. Replace memberlist implementation with github.com/grafana/memberlist

> After applying https://github.com/hashicorp/memberlist/commit/8d2a27af7d1d23119da66585d71f7b1e60fde3d3 and running TestSingleBinaryWithMemberlistScaling integ test, I am still seeing failed to unmarshal received KV Pair

That change is not enough. In github.com/grafana/memberlist we did more changes to properly fix it. You can diff github.com/grafana/memberlist with github.com/hashicorp/memberlist to see the actual differences.

pracucci commented:
> should we also add this behind some feature flag so it can be enabled in two steps?

That's a valid concern. I think we should add it behind feature flag to allow for a smoother migration.

> Btw, we have not see any OOM since we deployed this change on our load test environment, but we will keep it running for a couple of days more before we can confirm that this fixes https://github.com/cortexproject/cortex/issues/4668

We've experienced the same in the past when testing memberlist. It's caused by decoding of corrupted packets. It makes sense that this PR would fix it (or switching memberlist to github.com/grafana/memberlist).

alvinlin123 commented:
Thanks @pracucci After trying with `replace github.com/hashicorp/memberlist v0.2.4 => github.com/grafana/memberlist v0.2.5-0.20211201083710-c7bc8e9df94b`  The integration test does not print the unmarshal error anymore. Wether or not OOM issue persists is another story :) 

For release 1.12 I am thinking to just replace the hashcorp memberlist with grafana memberlist, because there was point in time where large message works, and we should keep it like that for release 1.12, I am less inclined to revert large message support for release 1.12. With that said, I am open to other thoughts.

pracucci commented:
I'm fine to move to `github.com/grafana/memberlist`: our OOM issues have been solved using it.

alvinlin123 commented on CHANGELOG.md:
```diff
@@ -3,8 +3,9 @@
 ## master / unreleased
 
 * [FEATURE] Ruler: Add `external_labels` option to tag all alerts with a given set of labels.
+* [FEATURE] Memberlist: Add `-memberlist.enable-broadcast-of-large-messages` option to enable broadcast of messages with more than 64KB.
```
We should release this as part of 1.12, so can you kindly: 

* Make this PR's target to be the release-1.12 branch instead of master; We will merge from release branch to master later on. 
* Put this line under the 1.12.0 change log. 

Thank you so much for working on this feature flag,

alvinlin123 commented on docs/configuration/config-file-reference.md:
```diff
@@ -3965,6 +3965,11 @@ The `memberlist_config` configures the Gossip memberlist.
 # CLI flag: -memberlist.message-history-buffer-bytes
 [message_history_buffer_bytes: <int> | default = 0]
 
+# Enable the broadcast of messages with more than 64KB. This can be safely
```
Do we need to briefly mention why one might not want to enable this "feature"?

marianafranco commented:
@alvinlin123 @pracucci I updated this PR to replace the memberlist implementation with github.com/grafana/memberlist and also added a new option to allow for smoother migrations.

pracucci commented:
Side note: I opened a PR to upstream the fix https://github.com/hashicorp/memberlist/pull/260.

pracucci commented:
> > should we also add this behind some feature flag so it can be enabled in two steps?
> 
> That's a valid concern. I think we should add it behind feature flag to allow for a smoother migration.

I think I said something wrong.

When I replied a couple of days ago I didn't remember the exact memberlist fix (I remember I fixed, but not the details). Today I opened a PR to upstream that fix, so I went through the code changes once again. The fix is only on the sender side, so I don't think we need any CLI flag to conditionally enable it.

marianafranco commented:
> The fix is only on the sender side, so I don't think we need any CLI flag to conditionally enable it.

Great! I created a new PR targeting the release-1.12 branch that only replaces the memberlist implementation: https://github.com/cortexproject/cortex/pull/4671
